doc/mgmt/SONiC_Design_Doc_Unified_FRR_Mgmt_Interface.md

```diff
@@ -0,0 +1,879 @@
+
+
+## 1.1 Requirements
+
+
+### 1.1.1 Functional Requirements
+
+  1. Extend Unified mode for full FRR-BGP config and management in SONiC
+  2. Extend sonic-cfggen, bgpcfgd and integrate with FRR-BGP for features supported in SONiC
+  3. Support for retrieval of FRR-BGP state and statistics information
+  4. For backward compatibility retain access to FRR UI (vtysh) for managing features that are NOT in conflict with SONiC features
+
+### 1.1.2 Configuration and Management Requirements
+
+1. Support Open Config data models for BGP config and Management
+2. Provide IS-CLI/gNMI/REST support for config and management of FRR-BGP features used in SONIC
+3. Enhance with Custom YANG models for features used in BGP that are not supported via Open Config data model
+4. Define ABNF schema for BGP features used in SONiC
+
+### 1.1.3 Scalability Requirements
+N/A
+
+### 1.1.4 Warm Boot Requirements
+As state and statistics information is retrieved from FRR-BGP on demand there is no Warm Boot specific requirements for this feature.
+
+## 1.2 Design Overview
+SONiC FRR-BGP Extended Unified config and management capability makes use of Management framework to implement the backend and transformer methods to support Open Config data models for BGP and route policy feature. The backend converts the incoming request to Redis ABNF schema format and writes the configuration to Redis DB. Then from DB events, bgpcfgd will configure FRR-BGP using FRR CLI commands.
+It also uses management framework's transformer methods to do syntactic and semantic validation of the requests using ABNF JSON before writing them into the Redis DB.
+
+### 1.2.1 Basic Approach
+
+* This enhancement takes comprehensive approach to support BGP features used in SONiC:
+	* Standard based YANG models and custom YANG models
+	* Open API spec
+	* Industry standard CLI
+	* Config Validation
+* REST server, gNMI server, Transformer methods - all in Go
+* Marshalling and unmarshalling using YGOT
+* Redis updated using CAS(Check-and-Set) trans. (No locking, No rollback)
+* Config Validation by using YANG model from ABNF schema
+
+### 1.2.2 Container
+
+There will be changes in following containers,
+* sonic-mgmt-framework
+* sonic-frr
+
+### 1.2.3 SAI Overview
+N/A - software feature
+
+
+# 2 Functionality
+## 2.1 Target Deployment Use Cases
+Configure and manage FRR-BGP via gNMI, REST and CLI interfaces using SONiC Management Framework.
+
+## 2.2 Functional Description
+Provide GNMI and REST support for config get/set, state get and statistics get,  CLI config and show commands for FRR-BGP features used in SONiC.
+
+
+# 3 Design
+## 3.1 Overview
+The extended unified config and management framework for FRR-BGP in SONiC is represented in below diagram.
+
+![FRR-BGP Unified Mgmt Framework](images/FRR-BGP-Unified-mgmt-frmwrk.png)
```

Comment:
The state and statistics flow in this diagram does not match with the CLI and REST get flow diagram. In the flow diagrams the response is through bgpcfgd but in this diagram it is shown that the FRR directly talks to management framework. Only one can be right.

Reply:
will fix this.

Reply:
Fixed the diagram, bgpcfgd is not involved for state/stats get.